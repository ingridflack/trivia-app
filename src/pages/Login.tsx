import {
  Box,
  Card,
  Container,
  Image,
  Link,
  Text,
  useToast,
} from "@chakra-ui/react";
import LoginForm from "../components/LoginForm";
import Header from "../components/Header";
import { useNavigate } from "react-router-dom";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../services/authService";
import useAuth from "../hooks/useAuth";

export interface LoginFormValues {
  email: string;
  password: string;
}

export default function Login() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormValues>();
  const navigate = useNavigate();
  const toast = useToast();
  const { saveUserData } = useAuth();

  const onSubmit: SubmitHandler<LoginFormValues> = async (values) => {
    try {
      const { data } = await AuthService.login(values);

      saveUserData(data);

      navigate("/");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  return (
    <>
      <Header />

      <Container
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection={["column", "column", "row"]}
          justifyContent="space-around"
          width="100%"
          padding="70px"
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Sign in
            </Text>
            <LoginForm
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              isSubmitting={isSubmitting}
              register={register}
            />

            <Box marginTop="20px" display="flex" justifyContent="space-between">
              <Link
                fontSize="small"
                color="gray.500"
                href="/trivia/recovery-email"
              >
                Forgot your password?
              </Link>

              <Link fontSize="small" color="gray.500" href="/sign-up">
                Create an account
              </Link>
            </Box>
          </Box>

          <Image
            src="src/assets/login.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
          />
        </Card>
      </Container>
    </>
  );
}
