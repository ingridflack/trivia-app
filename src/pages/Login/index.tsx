import { Box, Card, Container, Image, Text, useToast } from "@chakra-ui/react";
import LoginForm from "../../components/LoginForm";
import Header from "../../components/Header";
import { Link, useNavigate } from "react-router-dom";
import { SubmitHandler, useForm } from "react-hook-form";
import * as AuthService from "../../services/authService";
import * as TriviaService from "../../services/triviaService";
import useAuth from "../../hooks/useAuth";
import { Footer } from "../../components/Footer";

export interface LoginFormValues {
  email: string;
  password: string;
}

export default function Login() {
  const {
    handleSubmit,
    register,
    formState: { errors, isSubmitting },
  } = useForm<LoginFormValues>();
  const navigate = useNavigate();
  const toast = useToast();
  const { saveUserData } = useAuth();

  const onSubmit: SubmitHandler<LoginFormValues> = async (values) => {
    try {
      const { data } = await AuthService.login(values);

      saveUserData(data);

      const triviaConfig = localStorage.getItem("triviaConfig");

      if (triviaConfig) {
        const { data } = await TriviaService.createTrivia(
          JSON.parse(triviaConfig)
        );

        navigate(`/trivia/${data.triviaId}`);

        localStorage.removeItem("triviaConfig");
        return;
      }

      navigate("/");
    } catch (error) {
      console.log(error);
      toast({
        title: "An error occurred.",
        description: "Unable to login.",
        status: "error",
        duration: 2000,
        isClosable: true,
        position: "top-right",
      });
    }
  };

  return (
    <>
      <Header />

      <Container
        as="main"
        display="flex"
        alignItems="center"
        justifyContent="center"
        minH="calc(var(--chakra-vh) - 50px)" // 100vh - 50px (height of the header)
      >
        <Card
          display="flex"
          flexDirection="row"
          justifyContent="space-around"
          width="100%"
          padding={{ base: "20px", md: "70px" }}
          shadow="xl"
        >
          <Box maxW="392px" width="100%">
            <Text fontSize="xx-large" marginBottom="32px" color="gray.700">
              Sign in
            </Text>
            <LoginForm
              onSubmit={handleSubmit(onSubmit)}
              errors={errors}
              isSubmitting={isSubmitting}
              register={register}
            />

            <Box
              marginTop="20px"
              display="flex"
              justifyContent="space-between"
              fontSize="small"
              color="gray.500"
            >
              <Link to="/trivia/recovery-email">Forgot your password?</Link>

              <Link to="/sign-up">Create an account</Link>
            </Box>
          </Box>

          <Image
            src="src/assets/login.svg"
            alt="Login image"
            width="325px"
            height="324px"
            pointerEvents="none"
            display={{ base: "none", md: "none", lg: "block" }}
          />
        </Card>
      </Container>
      <Footer />
    </>
  );
}
