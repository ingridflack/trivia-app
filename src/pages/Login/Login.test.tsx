import { describe, expect, it } from "vitest";
import Login from ".";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

describe("Login", () => {
  it("should match the snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
