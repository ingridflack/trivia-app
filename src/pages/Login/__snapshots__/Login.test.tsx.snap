// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > should match the snapshot 1`] = `
<div>
  <header
    class="css-cyjglw"
  >
    <div
      class="chakra-container css-old1by"
    >
      <div
        class="chakra-stack css-2jbh02"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-kkaoep"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-j0ozid"
        >
          <a
            href="/login"
          >
             Sign in 
          </a>
          <a
            href="/sign-up"
          >
             Sign Up 
          </a>
        </div>
      </div>
    </div>
  </header>
  <main
    class="chakra-container css-3i1d8n"
  >
    <div
      class="chakra-card css-yp61un"
    >
      <div
        class="css-1updlva"
      >
        <p
          class="chakra-text css-mvfx24"
        >
          Sign in
        </p>
        <form
          class="chakra-stack css-2umoaz"
          role="form"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="email"
              id="field-:r0:-label"
            >
              Email address
            </label>
            <input
              class="chakra-input css-0"
              id="email"
              name="email"
              placeholder="Enter your email"
              type="email"
            />
          </div>
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="password"
              id="field-:r1:-label"
            >
              Password
            </label>
            <input
              class="chakra-input css-0"
              id="password"
              name="password"
              placeholder="Enter your password"
              type="password"
            />
          </div>
          <button
            class="chakra-button css-uxt1e8"
            type="submit"
          >
            Sign in
          </button>
        </form>
        <div
          class="css-1ddn805"
        >
          <a
            href="/trivia/recovery-email"
          >
            Forgot your password?
          </a>
          <a
            href="/sign-up"
          >
            Create an account
          </a>
        </div>
      </div>
      <img
        alt="Login image"
        class="chakra-image css-1b9ujuc"
        src="src/assets/login.svg"
      />
    </div>
  </main>
  <footer
    class="css-1s2lyti"
  >
    <div
      class="chakra-container css-0"
    >
      <div
        class="chakra-stack css-11h4xmw"
      >
        <a
          href="/"
        >
          <p
            class="chakra-text css-rekzoj"
          >
            Trivia
          </p>
        </a>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Site map
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/login"
            >
              Sign in
            </a>
            <a
              href="/sign-up"
            >
              Sign up
            </a>
          </div>
        </div>
        <div
          class="css-0"
        >
          <strong
            class="chakra-heading css-kf3i1p"
          >
            Social
          </strong>
          <div
            class="chakra-stack css-1ko7mmx"
          >
            <a
              href="https://www.linkedin.com/in/ingrid-flack/"
            >
              LinkedIn
            </a>
            <a
              href="https://github.com/ingridflack"
            >
              GitHub
            </a>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-1y3yl0q"
      >
        <p
          class="chakra-text css-0"
        >
          Â© 2024 Trivia App
        </p>
        <p
          class="chakra-text css-0"
        >
          Developed by Ingrid Flack
        </p>
      </div>
    </div>
  </footer>
</div>
`;
